Upgraded Cypress from 8.6.0 to 13.3.0

JasonLandbridge opened the pull request:
Hi there, 

## Motivation
The Cypress version is horribly outdated and I saw the intention here to upgrade: https://github.com/thefrontside/simulacrum/issues/234, so I went for it. 

There was also a bug I fixed where: 

```
const ClientPort = Number(process.env.PORT || 4000);
```
caused a `process is undefined` so I replaced that with a `Cypress.env('PORT')`, which works. I think `process` is not available anymore in the newer Cypress versions. 

## Approach

Major upgrade and migration of Cypress in the `@simulacrum/auth0-cypress` package

### Alternate Designs

 >   Explain what other alternatives you considered and why you chose this option.

 Not using this awesome `@simulacrum/auth0-cypress` package?

### Possible Drawbacks or Risks

Those updating @simulacrum/auth0-cypress will need to migrate their Cypress as well

### TODO in this PR
 - [ ] Update the `@simulacrum/auth0-cypress` [README](https://github.com/thefrontside/simulacrum/tree/v0/integrations/cypress#readme) with the new setup / usage process
 - [ ] Update the repo [README](https://github.com/thefrontside/simulacrum) with a reference to `@simulacrum/auth0-cypress`
 - [ ] Locally import into an [auth0/auth0-vue](https://github.com/auth0/auth0-vue ) and test it out, otherwise integrate
 - [ ] Ensure Auth0 cookies are persisted troughout tests in Cypress (see [Auth0 Cypress](https://docs.cypress.io/guides/end-to-end-testing/auth0-authentication))
 - [x] Look at upgrading all NPM packages in `@simulacrum/auth0-cypress`
       ![image](https://github.com/thefrontside/simulacrum/assets/15127381/abee8989-a7bb-4f2c-9088-83548ef5f495)

### Future TODOs 
  - [ ] Post this plugin at https://docs.cypress.io/plugins ([how to](https://github.com/cypress-io/cypress-documentation/blob/main/CONTRIBUTING.md#adding-plugins))
   

## Learning

 - https://docs.cypress.io/guides/references/migration-guide

## Screenshots

![image](https://github.com/thefrontside/simulacrum/assets/15127381/2b5bf2c6-618a-4c61-b2f3-bfdb4231c9d4)

dagda1 reviewed (approved):
Awesome work.  Thank you!  Most needed.

JasonLandbridge commented:
> Awesome work. Thank you! Most needed.

My pleasure! Could you please test it out and answer the questions when you have time? 

Thanks!

dagda1 commented:
> My pleasure! Could you please test it out and answer the questions when you have time?
@JasonLandbridge I have not worked with auth0 for a while now so my memory is a wee bit sketchy and I am a wee bit out of the loop with cookies.

The documentation is the README [here](https://github.com/thefrontside/simulacrum/tree/v0/integrations/cypress#readme).

I'm happy to let you drive whatever changes you feel appropriate and if I can help in any way, then please ask.

JasonLandbridge commented:
Thanks for the confidence :+1: 

I have updated the PR description with TODO's which I will work on Monday evening, I will keep you updated!

dagda1 commented:
> Thanks for the confidence :+1: 


Sir, I am thanking you.  This package needs a bit if love.  It was very useful when it was first created

dagda1 commented:
> Can you confirm Auth-simulator is working as is?

My guess is that you are starting the simulation server as the same PORT as the auth0 server.  They are actually 2 separate express servers.  We need the documentation clear on this.  I could not find much on mkcert in the docs but there is a very clear error message.

I've videoed myself starting the auth0 server 

First of all, I start the simulation server in the `packages/auth0` directory

```shell
PORT=4000 npm start
```

Then run the mutation with the auth0 port as 4400


Then I ran the mutation and all was good.

I've also included the subscription you can run to get more debug info.

Let me know 

https://github.com/thefrontside/simulacrum/assets/118328/6ba6b58d-1b58-4614-9654-29bb66645f4d

JasonLandbridge commented:
@dagda1 Thank you very much! I could indeed resolve my issue with your very clear instructions, I will continue working on this PR this weekend

dagda1 commented:
> @dagda1 Thank you very much! I could indeed resolve my issue with your very clear instructions, I will continue working on this PR this weekend

Don't be afraid to ask again.  I don't think the instructions on how to use the auth0 simulator are clear and that has harmed adoption.

I believe this tool could be very useful and there is nothing else like it that I know of.

JasonLandbridge commented:
> Don't be afraid to ask again. I don't think the instructions on how to use the auth0 simulator are clear and that has harmed adoption.

Thanks, I'm updating the README as I figure out the things that were unclear for me. And I would be more than happy to post this plugin at https://docs.cypress.io/plugins ([how to](https://github.com/cypress-io/cypress-documentation/blob/main/CONTRIBUTING.md#adding-plugins)). This was the first place I looked, and there was unfortunately nothing that did what I needed. Then I came across [the article](https://frontside.com/blog/2022-01-13-auth0-simulator/) describing the concept and here I am. 

> I believe this tool could be very useful and there is nothing else like it that I know of.

Definitely, I searched as well and couldn't find anything like it and it can be revived. Like an old car that needs some maintenance before it can be used again!

So as of now I have it working again and both example projects are working with everything upgraded in this PR. There is a minor bug in the `nextjs-auth0` example project where in the console it returns 401's for `http://localhost:3000/api/auth/me 401 (Unauthorized)`. I think it might be related to the packages being outdated in the example projects but I will make a separate PR for that. 

Steps for now:
1. Cleanup this PR
2. Create a Vue.js/Vite example project and add Auth0-Vue integration in a separate PR
3. Test everything with a real life application
4. Upgrade the other example projects
5. Go through README's

JasonLandbridge commented:
@dagda1 It's done! :tada:  
I changed, fixed and refactored a lot of things but it's now fully working with all 3 example apps and every package is now upgraded as well. I've updated the README with the new way of doing things, most breaking changes will come from upgrading Cypress anyway. Fixed many bugs, probably added a few new ones, so I hope if you have time that you can check it out.

Tommorrow (Sunday), I will create a new PR branching from this one where I will do the Vue/vite integration with example projects :+1: 

Cheers!

dagda1 commented:
I think this is great. Excellent work

@jbolda does this project still use covector to publish?

JasonLandbridge commented:
> I think this is great. Excellent work
> 
> @jbolda does this project still use covector to publish?

Thanks! Today I got a chance to test this project at work and the packaging is I think not working correctly (I'm linking it locally) due to the project structure change, but I'm still investigating why it's not working. This PR can be merged and then we can continue in the other #271 PR where I did the Vue integration and am submitting my fixes. Then a new version would be greatly appreciated once that PR is merged. After which, I will make separate bug fixes for any bugs I may find :+1:

jbolda commented:
@dagda1 aye, `npx covector add` (or equivalent package manager) to add a change file.

dagda1 commented on integrations/cypress/README.md:
```diff
@@ -1,17 +1,32 @@
 # @simulacrum/auth0-cypress
 
-[Cypress](https://www.cypress.io/) [Auth0](https://auth0.com) addon that simulates an auth0 server running on `localhost` so you don't have to create fake accounts while developing or running tests that require authentication.
+A [Cypress](https://www.cypress.io/) plugin specifically created for SDK's that integrate with [Auth0](https://auth0.com)! @simulacrum/auth0-cypress will simulate an auth0 server, running on `localhost`, so you don't have to create fake Auth0 accounts while developing or running tests that require authentication!
```
> 1. Is the `mkcert` command needed to make the encrypt task work when configuring Cypress?

The encrypt task used to be needed to create the auth0 cookie.  I think you no longer do that.  Why is that.

> 2. Is the `mkcert` needed because the Auth0Simulation server runs on an https localhost domain?

Yes

> 3. "Should" this Cypress plugin work when running in a CI/CD pipeline? Maybe add some instructions for that as well if it should work?

Not something I've ever done

> 4. Is it maybe better to add this as a step 6 under "Installation" when setting everything up instead of the project description?

That seems better

JasonLandbridge replied:
> The encrypt task used to be needed to create the auth0 cookie. I think you no longer do that. Why is that.

I have some trouble remembering, but I think I removed it due to the command, in how it was used here, being [deprecated](https://docs.cypress.io/api/cypress-api/cookies) (see history at the bottom) but I might not have properly replaced it with `cy.origin()` or `cy.session()`. That is also something on my todo list on how that should work in the newer Cypress versions

I left the encrypt task in it, but it indeed might be removable due to the migration. Is it okay if I create some issues with these points which I will pick up in separate PR's to keep it better organized?

Also for the `mkcert` point that the README should reflect it better

JasonLandbridge replied:
I have added a description with a link as a step 6 to the README :+1:

dagda1 commented on integrations/cypress/cypress/support/commands/index.ts:
```diff
@@ -1,45 +1,52 @@
-Cypress.Commands.add('destroySimulation', makeDestroySimulation({ atom, getClientFromSpec }));
+    if (provider === 'auth0_react') {
+        registerAuth0ReactCommands();
+    }
```
I think there should be an assertion of default values if there is no `sdk` value or there will be no login command registered and nothing will work.

JasonLandbridge replied:
I've added a throw error here

dagda1 commented on integrations/cypress/cypress/support/utils/config.ts:
```diff
@@ -10,35 +10,36 @@ interface Config {
-    domain,
-    sdk
+    auth0RunningPort,
+    auth0SimulatorPort: Cypress.env('AUTH0_SIMULATOR_PORT') ?? 4000,
```
Do we need this, you can set this using `PORT=4000` without another environment variable.

JasonLandbridge replied:
PORT is quite generic and it was initially confusing for me that this was not actual Auth0 server but the GraphQL interface so I chose to make it distinct from each-other.

JasonLandbridge replied:
After re-reading this I understand it now, the port can already be changed so this extra `env` variable is not needed. I ended up removing it altogether

dagda1 replied:
I managed to get your tests passing in `./examples/create-react-app`.  There were a few gotchas and especially running `--headless`.

Any `cy.log` in a promise eg. `cy.log(`successfully logged in with token ${JSON.stringify(token)}`);` will cause the test to fail with 

>  CypressError: The following error originated from your test code, not from Cypress. It was caused by an unhandled promise rejection.

  > Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise.

The command that returned the promise was:

  > `cy.visit()`

The cy command you invoked inside the promise was:

  > `cy.log()`

JasonLandbridge replied:
Any chance this PR can be merged, and not yet publish a new package, and then we continue from the other PR? The other PR already include fixes for these and I fixed the packaging in that one. It has become difficult to track what is what a new issues keep arising. 

Or I merge both into 1 but then the PR is even more huge

dagda1 replied:
> Any chance this PR can be merged, and not yet publish a new package, and then we continue from the other PR? The other PR already include fixes for these and I fixed the packaging in that one. It has become difficult to track what is what a new issues keep arising.
> 
> Or I merge both into 1 but then the PR is even more huge

I personally would not merge this because the `--headless` tests are not working because the calls to `cy.log` make them fail. 

Also the example tests in the create-react-app project do not work and have not been updated.  I quickly got a [hacky branch](https://github.com/thefrontside/simulacrum/tree/pc/create-react-app-fixes) together to get them passing.

That branch was a total hack

That is my shout anyway, but I do not work for frontside anymore, so maybe @taras @jbolda or @cowboyd have a different opinion.  I'm also a bit pushed for time at the moment

I really appreciate the work you are doing but I don't like merging stuff that is not finished but the others might differ.

JasonLandbridge replied:
Fair points, I underestimated the number of changes that were needed and at this point it's almost a complete rewrite. Ideally several smaller PR's would have been in order. 

Thanks for the branch, I will take a look and use it to fix this PR!

dagda1 commented on integrations/cypress/cypress/support/utils/config.ts:
```diff
@@ -10,35 +10,36 @@ interface Config {
-  let domain = Cypress.env('domain') ?? 'localhost:4400';
-  let sdk = (Cypress.env('AUTH0_SDK') ?? 'auth0_react') as Auth0SDKs;
-
+  let auth0RunningPort = Cypress.env('AUTH0_RUNNING_PORT') ?? 4400;
```
```suggestion
  let auth0RunningPort = Cypress.env('AUTH0_SIMULATOR_PORT') ?? 4400;
```

I think we only need 1 of these variables and callit `AUTH0_SIMULATOR_PORT`

JasonLandbridge replied:
Sure, i think there won't be a reason a user would need to change this port. I will update it :+1:

JasonLandbridge replied:
I have fixed this :+1:

JasonLandbridge commented:
Due to the high necessity of this PR for my work, and the overall huge chaos this PR has become due to my own doing, I have decided to create my own [JasonLandbridge/Cypress-Auth0-Simulator](https://github.com/JasonLandbridge/Cypress-Auth0-Simulator) with all the changes where I will continue the work. I will work towards removing everything except the plugin and examples. Then I can publish it as a NPM package and will keep maintaining that.  

It is not my intention to steal all the work everyone here has done, so I will make sure to reference this repo and credit where I can. I did flatten the repo only so I can start from a "blank slate" but I do intend to keep using the simulacrum packages. It will basically be only do the Cypress integration plugin part. 

@dagda1, Thank you very much for your time and help! I'm closing this PR now to avoid taking up more of your time in going through this PR.
